### PR TITLE
[MXNET-86] Revert to pre-profile-changes copy code

### PR DESCRIPTION
**Pull request by cjolivier01**
## Description ##

Fix for: https://github.com/apache/incubator-mxnet/issues/10042

For whatever reason, the new approach isn't fork-friendly.  Just reverted to the previous code for this function from before my PR: 106f97f1881e6bb1a00c56a0ae55200e27297733

The code change in that PR was just a small performance-optimization change.

### Essentials ###
- [ ] Passed code style checking (`make lint`)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by marcoabreu** [comment]
Would you mind creating a test with the MVE the user provided to ensure this does not happen again?

File: src/ndarray/ndarray_function.cc [hunk 1]
```diff
@@ -26,35 +26,24 @@
...
   MSHADOW_TYPE_SWITCH(to->type_flag_, DType, {
     if (to->type_flag_ == from.type_flag_) {
-      TBlob dest = to->FlatTo1D<cpu, DType>();
-      TBlob src = from.FlatTo1D<cpu, DType>();
-      const size_t size = src.Size();
-      if (dest.CheckContiguous() && src.CheckContiguous() && size >= 20000 /* non-trivial size */) {
-        CHECK_EQ(dest.shape_, src.shape_)
-          << "Copy:shape mismatch:" << dest.shape_ << " vs " << src.shape_;
-          mxnet_op::Kernel<mxnet_op::op_with_req<mshadow_op::identity, kWriteTo>, cpu>::Launch(
-            ctx.get_stream<cpu>(), src.Size(), dest.dptr<DType>(), src.dptr<DType>());
```
Comment (sxjscience) [inline src/ndarray/ndarray_function.cc]: So we cannot use kernel::launch together with shared memory?

Comment (cjolivier01) [reply src/ndarray/ndarray_function.cc]: I don't see why not, but I am just going to revert now because the change wasn't worth the trouble it caused.

Comment (cjolivier01) [reply src/ndarray/ndarray_function.cc]: It does get stuck in the kernel launch code for whatever reason.

Comment (sxjscience) [reply src/ndarray/ndarray_function.cc]: Yes, I think we can revert first and find the cause later. What do you think @marcoabreu . This error is really critical as the DataLoader is really common among users.

Comment (cjolivier01) [reply src/ndarray/ndarray_function.cc]: I think the fork doesn't play nicely with libgomp in this codepath because it works fine when linked against mkl, which uses libiomp.  Or if built with cmake, which builds and links against Intel omp (in 3rdparty), it works fine.

Comment (sxjscience) [reply src/ndarray/ndarray_function.cc]: Using the keyword I find this link: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=58378

Comment (cjolivier01) [reply src/ndarray/ndarray_function.cc]: Ha! Nice find! That's exactly it.

Comment (sxjscience) [reply src/ndarray/ndarray_function.cc]: It suggests start the multiprocessing using `forkserver` (https://docs.python.org/3/library/multiprocessing.html#multiprocessing.set_start_method). I'm trying this keyword.

File: tests/python/unittest/test_gluon_data.py [hunk 2]
```diff
@@ -112,6 +117,76 @@ def test_multi_worker():
...
+                labels[i] = np.pad(labels[i], (0, max_labels_len - labels[i].shape[0]),
+                                   'constant', constant_values=-1)
+
+            inputs = np.asarray(inputs, dtype=np.float32)
+            if labels is not None:
+                labels = np.asarray(labels, dtype=np.float32)
+            inputs = inputs.transpose((1, 0, 2))
+            labels = labels.transpose((1, 0))
+
+            return (nd.array(inputs, dtype=inputs.dtype, ctx=context.Context('cpu_shared', 0)),
```
Comment (marcoabreu) [inline tests/python/unittest/test_gluon_data.py]: I guess the context is not really important here, right?

File: src/ndarray/ndarray_function.cc (hunk 1 above)
Comment (cjolivier01) [reply src/ndarray/ndarray_function.cc]: Intel OMP at fork, sets all its init vars to uninitialized, so that I suppose it recreates its state in the new fork:

https://github.com/pmodels/bolt/blob/master/runtime/src/z_Linux_util.c#L1536

File: tests/python/unittest/test_gluon_data.py [hunk 3]
```diff
@@ -112,6 +117,76 @@ def test_multi_worker():
...
+                    nd.array(x_lens, ctx=context.Context('cpu_shared', 0))) \
+                if labels is None else (
+                nd.array(inputs, dtype=inputs.dtype, ctx=context.Context('cpu_shared', 0)),
+                nd.array(x_lens, ctx=context.Context('cpu_shared', 0)),
+                nd.array(labels, dtype=labels.dtype, ctx=context.Context('cpu_shared', 0)),
+                nd.array(y_lens, ctx=context.Context('cpu_shared', 0)))
+
+
+    data = Dummy(True)
+    loader = DataLoader(data, batch_size=40, batchify_fn=data.batchify, num_workers=2)
```
Comment (marcoabreu) [inline tests/python/unittest/test_gluon_data.py]: Could you add a comment that explains that we basically just expect this to pass?

File: tests/python/unittest/test_gluon_data.py (hunk 2 above)
Comment (Jerryzcn) [reply tests/python/unittest/test_gluon_data.py]: might be related to 'cpu_shared'.

Comment (cjolivier01) [reply tests/python/unittest/test_gluon_data.py]: Why would you think that?

Comment (Jerryzcn) [reply tests/python/unittest/test_gluon_data.py]: when I trained on our dataset, the stack trace point to cpu_shared_storage_manager.h:178.
However, it might not be, need to be tested.

File: tests/python/unittest/test_gluon_data.py (hunk 3 above)
Comment (cjolivier01) [reply tests/python/unittest/test_gluon_data.py]: ok

File: src/ndarray/ndarray_function.cc (hunk 1 above)
Comment (cjolivier01) [reply src/ndarray/ndarray_function.cc]: Looks like someone tried to fix it once: https://patchwork.ozlabs.org/patch/319827/

File: tests/python/unittest/test_gluon_data.py (hunk 2 above)
Comment (cjolivier01) [reply tests/python/unittest/test_gluon_data.py]: I meant Marco's comment, I didn't see yours yet at the time.  Since they are foked, any memory allocated after the fork would need to be cpu_shared.

File: src/ndarray/ndarray_function.cc (hunk 1 above)
Comment (sxjscience) [reply src/ndarray/ndarray_function.cc]: Failed to use `forkserver`. I'll continue working on something else.

Comment (marcoabreu) [reply src/ndarray/ndarray_function.cc]: @sxjscience sure, go ahead and merge. This seems pretty critical

File: tests/python/unittest/test_gluon_data.py (hunk 2 above)
Comment (marcoabreu) [reply tests/python/unittest/test_gluon_data.py]: My thought was that the internals of the DataLoader are being tested here and that they are independent of the context. Am I mistaken here?

Comment (marcoabreu) [reply tests/python/unittest/test_gluon_data.py]: Is the logic here that it loads the Data multithreaded into CPU before it gets passed down further (cpu/gpu context)?

Comment (sxjscience) [reply tests/python/unittest/test_gluon_data.py]: @marcoabreu In fact, the current design of DataLoader requires the user to correctly define the batchify function in order to use the multi-process feature. We need to revise the interface.

Comment (marcoabreu) [reply tests/python/unittest/test_gluon_data.py]: Ah I see, thanks

Comment (cjolivier01) [reply tests/python/unittest/test_gluon_data.py]: No, what's being tested is the fork()

**Comment by sxjscience** [comment]
@cjolivier01 Eric said that `pthread_atfork` http://pubs.opengroup.org/onlinepubs/007904975/functions/pthread_atfork.html might be helpful for solving the problem.

**Comment by cjolivier01** [comment]
yeah I actually tried setting omp_num_threads() to one there earlier today but it didn’t fix it. I think we’ll need to find proc addresses for some gomp functions like in that link above and reset libgomp.  Be cool if someone can create a JIRA ticket for that
